Version v11.5.1

danjm opened the pull request.

danjm committed:
Only call _loginUser if there is a password in env variables (#21854)

Fixes #21598
Fixes #21837
Fixes #21834

There is a bug in production that prevents users from opening MetaMask
if when they last stopped MetaMask (usually by closing the browser), the
network they had selected is no longer working. In these cases, when
they attempt to open MetaMask, there is an infinite spinner.

What is happening is that [await
this.accountTracker._updateAccounts();](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L3408)
in `_loginUser` is never resolving, so `this._startUISync();` is never
called
```
async _loginUser() {
    try {
      // Automatic login via config password
      const password = process.env.PASSWORD;
      if (password && !process.env.IN_TEST) {
        await this.submitPassword(password);
      }

      // Updating accounts in this.accountTracker before starting UI syncing ensure that
      // state has account balance before it is synced with UI
      await this.accountTracker._updateAccounts();
    } finally {
      this._startUISync();
    }
  }
```

which means that
[this](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/ui.js#L200C18-L200C18)
`initializeUiWithTab(activeTab);` call in `ui.js` never occurs, and so
the UI is never initialized.

The `await this.accountTracker._updateAccounts();` is not resolving
because ` const balances = await ethContract.balances(addresses,
ethBalance);` in `_updateAccountsViaBalanceChecker` in
`account-tracker.js` never resolves.

This is happening because calling `ethContract.balances` results in a
`eth_call` request being passed through our middleware, and it has to go
through the `createBlockRefRewriteMiddleware`, [which has the
following](https://github.com/MetaMask/eth-json-rpc-middleware/blob/main/src/block-ref-rewrite.ts#L41):
`const latestBlockNumber = await blockTracker.getLatestBlock();`

The `blockTracker.getLatestBlock` function has this line of code:
```
    const latestBlock: string = await new Promise((resolve) =>
      this.once('latest', resolve),
    );
```

So ultimately our code is waiting on a `'latest'` event to be emitted
from within a block tracker. But because the user is on a network that
is broken, no successful block polling is occuring, and therefore a
`'latest'` block/event is never emitted, and therefore `getLatestBlock`
is awaiting forever, and so-on up the call stack.

The behaviour of the blocktracker is long standing. However, users have
begun to experience this bug recently because of this commit
https://github.com/MetaMask/metamask-extension/commit/48184febedd6b39621ce305ae29089e208ccb0bc

That commit introduced the listener for `startUISync` in `ui.js`, which
is what is not occuring when we get stuck on the above mentioned
`getLatestBlock` call. That commit was introduced to prod with v11.3.0

To quickly fix this bug so that we can release a hotfix, this PR
eliminates the `_loginUser` call in the MetaMask Controller when we are
not in a local dev build. It makes that call dependent on the presence
of `process.env.PASSWORD`, which only occurs on development builds. This
is acceptable because `_loginUser` is only used for the purpose of
allowing devs to skip the password screen when doing local development.
It does not serve any useful purpose in production.

A deeper fix would possibly be to introduce a timeout at some point in
the call stack, or to have `getLatestBlock` also listen for some sort of
error event, and throw up an error if it occurs.

Here is how I reproduce this bug on `develop`:

1. Build metamask locally from the develop branch
2. Copy the contents of `dist/chrome` to a new folder, and install
metamask from this folder. Onboard.
3. Go to chainlist.org, toggle open the Ethereum Mainnet options, and
add one of these eth mainnet custom networks, and switch to that network
4. Open the background console
5. Enter `chrome.storage.local.get(console.log)` into the console. An
object will be logged. Inspect the object and in particular the
NetworkController. Find the unique id of the network you just added. It
will look something like `0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e`
6. For the below script, replace each instance of
`0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e` with the unique network id you
copied in step 3. Copy and paste this revised script to the background
console and run it
7. Reload the metamask-extension
8. Attempt to open the extension and you should see an infinite spinner,
with a white background and no log in screen. Reloading the extension
and closing an re-opening the browser doesn't fix the problem.

Then, switch to the branch of this PR, build locally, and copy the new
contents of `dist/chrome` to the folder you created in step 2 and reload
the extension. Now you should still see a loading spinner, but you
should be able to switch networks and then login

The script to modify and run in step 6 is:
```
chrome.storage.local.get(({ data, meta }) => {
    let currentData = structuredClone(data)

    let newNetworkConfigurations = {
        ...currentData.NetworkController.networkConfigurations
    }
    newNetworkConfigurations["0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e"] = {
        "chainId": "0x1",
        "id": "0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e",
        "nickname": "Pokt",
        "rpcPrefs": {},
        "rpcUrl": "https://eth-rpc.gateway.pokt.network",
        "ticker": "ETH"
    }

    let newProviderConfig = {
        ...currentData.NetworkController.providerConfig
    }
    newProviderConfig["0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e"] = {
        "chainId": "0x1",
        "id": "0b29f8c9-bc0e-4058-a4a1-c39c5a5bb90e",
        "nickname": "Pokt",
        "rpcPrefs": {},
        "rpcUrl": "https://eth-rpc.gateway.pokt.network",
        "ticker": "ETH",
        "type": "rpc"
    }

    let newData = {
        ...currentData,
        NetworkController: {
            ...currentData.NetworkController,
            networkConfigurations: newNetworkConfigurations,
            providerConfig: newProviderConfig,
        }

    }

    return chrome.storage.local.set({
        data: newData,
        meta,
    }, () => { window.location.reload() })
})
```

![Screenshot from 2023-11-15
16-20-00](https://github.com/MetaMask/metamask-extension/assets/7499938/84b8596c-1ebc-481f-8f28-d4a24e96752e)

![Screenshot from 2023-11-15
16-15-42](https://github.com/MetaMask/metamask-extension/assets/7499938/1a327fd6-5101-4aad-b3a4-7704d50ffcdb)

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

danjm committed:
v11.5.1

pedronfigueiredo commented on app/scripts/metamask-controller.js:
```diff
@@ -3374,17 +3377,11 @@ export default class MetamaskController extends EventEmitter {
   }
 
-  async _loginUser() {
+  async _loginUser(password) {
```
Do we want to add this argument to the function call as well?

danjm replied:
done with most recent cherry-pick

zone-live committed:
chore: pass the password into method _loginUser (#21863)

## **Description**

Auto login stopped working.

**Fix:**
Pass the `password` into `_loginUser`.

## **Related issues**

Fixes: #

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.
